src/Autofac/Builder/RegistrationBuilder.cs

```diff
@@ -61,6 +62,12 @@ public static class RegistrationBuilder
     public static IRegistrationBuilder<TImplementer, ConcreteReflectionActivatorData, SingleRegistrationStyle> ForType<TImplementer>()
         where TImplementer : notnull
     {
+        // Open generics can't be generic type parameters so we don't have to check for that here.
```

Comment:
`RegistrationBuilder.ForType` (both overloads) is where all the `RegisterType<T>` and `RegisterType(Type t)` calls come through. It also gets used a few places directly like in ACTNARS and assembly/type scanning.

Adding the check here means we can catch any issues where someone is trying to do something that probably won't work with reflection activation and provide a more proactive error.

It does mean, though, that we may do the same checks twice on a given type. For example, _while scanning the set of types_ to register, we'll filter off things that don't match this criteria; then when the registration gets created it will be checked _again_ here. Ideally it means we won't get any of these errors during any sort of scanning, but it may have container build performance impacts due to the double-check.